sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/graph/FlowAnalysis.scala

```diff
@@ -0,0 +1,316 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.pipelines.graph
+
+import scala.util.Try
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.{AliasIdentifier, TableIdentifier}
+import org.apache.spark.sql.catalyst.analysis.{CTESubstitution, UnresolvedRelation}
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, SubqueryAlias}
+import org.apache.spark.sql.classic.{DataFrame, Dataset, DataStreamReader, SparkSession}
+import org.apache.spark.sql.pipelines.{AnalysisWarning, Sql}
+import org.apache.spark.sql.pipelines.graph.GraphIdentifierManager.{
+  ExternalDatasetIdentifier,
+  InternalDatasetIdentifier
+}
+import org.apache.spark.sql.pipelines.util.{
+  BatchReadOptions,
+  InputReadOptions,
+  StreamingReadOptions
+}
+
+object FlowAnalysis {
+  def createFlowFunctionFromLogicalPlan(plan: LogicalPlan): FlowFunction = {
+    new FlowFunction {
+      override def call(
+          allInputs: Set[TableIdentifier],
+          availableInputs: Seq[Input],
+          confs: Map[String, String],
+          currentCatalog: Option[String],
+          currentDatabase: Option[String]
+      ): FlowFunctionResult = {
+        val ctx = FlowAnalysisContext(
+          allInputs = allInputs,
+          availableInputs = availableInputs,
+          currentCatalog = currentCatalog,
+          currentDatabase = currentDatabase,
+          spark = SparkSession.active
+        )
+        val df = try {
+          confs.foreach { case (k, v) => ctx.setConf(k, v) }
+          Try(FlowAnalysis.analyze(ctx, plan))
+        } finally {
+          ctx.restoreOriginalConf()
+        }
+        FlowFunctionResult(
+          requestedInputs = ctx.requestedInputs.toSet,
+          usedBatchInputs = ctx.batchInputs.toSet,
+          usedStreamingInputs = ctx.streamingInputs.toSet,
+          usedExternalInputs = ctx.externalInputs.toSet,
+          dataFrame = df,
+          sqlConf = confs,
+          analysisWarnings = ctx.analysisWarnings.toList
+        )
+      }
+    }
+  }
+
+  /**
+   * Constructs an analyzed [[DataFrame]] from a [[LogicalPlan]] by resolving Pipelines specific
+   * TVFs and datasets that cannot be resolved directly by Catalyst. This is used in SQL pipelines
+   * and Python pipelines (when users call spark.sql).
+   *
+   * This function shouldn't call any singleton as it will break concurrent access to graph
+   * analysis; or any thread local variables as graph analysis and this function will use
+   * different threads in python repl.
+   *
+   * @param plan     The [[LogicalPlan]] defining a flow.
+   * @param forAnalysisApi Whether the query is being analyzed through Analysis API algorithm.
+   * @return An analyzed [[DataFrame]].
+   */
+  private def analyze(
+      context: FlowAnalysisContext,
+      plan: LogicalPlan,
+      forAnalysisApi: Boolean = false
```

Comment:
I think we can remove the `forAnalysisApi` argument (and associated docstring) – it's not set to `true` anywhere.